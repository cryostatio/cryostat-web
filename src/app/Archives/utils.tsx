/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RecordingDirectory, Target } from '@app/Shared/Services/api.types';

export const includesDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): boolean => {
  return arr.some((t) => t.connectUrl === dir.connectUrl);
};

export const indexOfDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): number => {
  let index = -1;
  arr.forEach((d, idx) => {
    if (d.connectUrl === dir.connectUrl) {
      index = idx;
    }
  });
  return index;
};

export const getTargetFromDirectory = (dir: RecordingDirectory): Target => {
  return {
    connectUrl: dir.connectUrl,
    alias: dir.jvmId,
    labels: [],
    annotations: {
      cryostat: [],
      platform: [],
    },
  };
};
