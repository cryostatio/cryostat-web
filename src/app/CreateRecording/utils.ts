/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { TemplateType } from '@app/Shared/Services/api.types';
import { EventTemplateIdentifier } from './types';

export const RecordingNamePattern = /^[\w_]+$/;

export const DurationPattern = /^[1-9][0-9]*$/;

export const isRecordingNameValid = (name: string) => RecordingNamePattern.test(name);

export const isDurationValid = (duration: number) => DurationPattern.test(`${duration}`);

export const templateFromEventSpecifier = (eventSpecifier: string): EventTemplateIdentifier | undefined => {
  const regex = /^template=([a-zA-Z0-9]+)(?:,type=([a-zA-Z0-9]+))?$/im;
  if (eventSpecifier && regex.test(eventSpecifier)) {
    const parts = regex.exec(eventSpecifier);
    if (parts) {
      return {
        name: parts[1],
        type: parts[2] as TemplateType,
      };
    }
  }
  return undefined;
};
