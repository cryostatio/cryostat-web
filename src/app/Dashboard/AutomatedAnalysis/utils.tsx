/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AnalysisResult, Evaluation } from '@app/Shared/Services/api.types';
import { Stack, StackItem, Label, Title, Text } from '@patternfly/react-core';
import _ from 'lodash';

export const transformAADescription = (result: AnalysisResult): JSX.Element => {
  const format = (s: Evaluation): JSX.Element => {
    if (typeof s === 'string') {
      return <Text>{s}</Text>;
    }
    if (Array.isArray(s)) {
      return (
        <Stack>
          {s.map((e) => (
            <StackItem key={e.setting}>
              <Title headingLevel={'h6'}>{e.setting}</Title>
              <Label>
                {e.name}={e.value}
              </Label>
            </StackItem>
          ))}
        </Stack>
      );
    }
    throw `Unrecognized item: ${s}`;
  };
  return (
    <div>
      {Object.entries(result.evaluation || {}).map(([k, v]) =>
        v && v.length ? (
          <div key={k}>
            <span>
              <Title headingLevel={'h5'}>{_.capitalize(k)}</Title>
              {format(result.evaluation[k])}
            </span>
            <br />
          </div>
        ) : (
          <div key={k}></div>
        ),
      )}
    </div>
  );
};
