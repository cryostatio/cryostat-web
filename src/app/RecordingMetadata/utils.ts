/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';
import { Observable, from } from 'rxjs';

export const isEqualLabel = (a: KeyValue, b: KeyValue) => {
  return a.key === b.key && a.value === b.value;
};

export const includesLabel = (arr: KeyValue[], searchLabel: KeyValue) => {
  return arr.some((l) => isEqualLabel(searchLabel, l));
};

export const parseLabelsFromFile = (file: File): Observable<KeyValue[]> => {
  return from(
    file
      .text()
      .then(JSON.parse)
      .then((obj) => {
        const labels: KeyValue[] = [];
        const labelObj = obj['labels'];
        if (labelObj) {
          Object.keys(labelObj).forEach((key) => {
            labels.push({
              key: key,
              value: labelObj[key],
            });
          });
          return labels;
        }
        throw new Error('No labels found in file');
      }),
  );
};

export const getLabelDisplay = (label: KeyValue) => `${label.key}:${label.value}`;

export const LabelPattern = /^\S+$/;

export const getValidatedOption = (isValid: boolean) => {
  return isValid ? ValidatedOptions.success : ValidatedOptions.error;
};

export const matchesLabelSyntax = (l: KeyValue) => {
  return l && LabelPattern.test(l.key) && LabelPattern.test(l.value);
};
