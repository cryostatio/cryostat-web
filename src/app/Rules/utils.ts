/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ruleObjKeys } from './types';

export const isRule = (obj: object): boolean => {
  for (const key of ruleObjKeys) {
    if (!Object.prototype.hasOwnProperty.call(obj, key)) {
      return false;
    }
  } // Ignore unknown fields
  return true;
};

// FIXME check if this is correct/matches backend name validation
export const RuleNamePattern = /^[\w_]+$/;

export const isRuleNameValid = (name: string) => RuleNamePattern.test(name);

export const CEL_SPEC_HREF = 'https://github.com/google/cel-spec/blob/master/doc/langdef.md';
