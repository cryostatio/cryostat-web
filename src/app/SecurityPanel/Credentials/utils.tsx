/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import { AuthCredential } from '@app/AppLayout/CredentialAuthForm';
import { StreamOf } from '@app/utils/utils';
import * as React from 'react';
import { debounceTime, Subscription } from 'rxjs';

export const CredentialContext = React.createContext(new StreamOf<AuthCredential>({ username: '', password: '' }));

export interface TestRequest {
  id: string;
  targetUrl: string;
  data?: unknown;
}

// Each test request registers itself to test pool when initiated. When completed, remove itself from pool.
// Auth form will poll this pool for a set time to determine if form should is disabled.
export const TestPoolContext = React.createContext(new Set<TestRequest>());

export const useAuthCredential = (
  ignoreEmit?: boolean,
  debounceMs = 50
): [AuthCredential, (credential: AuthCredential) => void] => {
  const [credential$, setCredential$] = React.useState<AuthCredential>({ username: '', password: '' });
  const authCredentialContext = React.useContext(CredentialContext);

  React.useEffect(() => {
    let sub: Subscription | undefined;
    if (!ignoreEmit) {
      sub = authCredentialContext.get().pipe(debounceTime(debounceMs)).subscribe(setCredential$);
    }
    return () => sub && sub.unsubscribe();
  }, [setCredential$, authCredentialContext, debounceMs, ignoreEmit]);

  const setCredential = React.useCallback(
    (credential: AuthCredential) => {
      authCredentialContext.set(credential);
    },
    [authCredentialContext]
  );

  return [credential$, setCredential];
};
