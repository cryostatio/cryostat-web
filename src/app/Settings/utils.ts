/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FeatureLevel } from '@app/Shared/Services/service.types';
import { hashCode } from '@app/utils/utils';
import { SettingTab, ThemeSetting, _TransformedUserSetting } from './types';

export const tabAsParam = (key: SettingTab) => {
  const parts = key.split('.');
  return parts[parts.length - 1].toLowerCase().replace(/[_]/g, '-');
};

export const paramAsTab = (param: string) => {
  return `SETTINGS.CATEGORIES.${param.toUpperCase().replace(/[-]/g, '_')}`;
};

export const selectTab = (tabKey: SettingTab) => {
  const tab = document.getElementById(`pf-tab-${tabKey}-${hashCode(tabKey)}`);
  tab && tab.click();
};

export const getDefaultTheme = (): ThemeSetting => {
  if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
    return ThemeSetting.DARK;
  }
  return ThemeSetting.LIGHT;
};

export const getGroupFeatureLevel = (settings: _TransformedUserSetting[]): FeatureLevel => {
  if (!settings.length) {
    return FeatureLevel.DEVELOPMENT;
  }
  return settings.slice().sort((a, b) => b.featureLevel - a.featureLevel)[0].featureLevel;
};
