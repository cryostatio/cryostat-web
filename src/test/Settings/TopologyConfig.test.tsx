/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { TopologyConfig } from '@app/Settings/Config/TopologyConfig';
import { defaultServices } from '@app/Shared/Services/Services';
import { cleanup } from '@testing-library/react';
import React from 'react';
import { of } from 'rxjs';
import { renderSnapshot } from '../utils';

jest.spyOn(defaultServices.api, 'doGet').mockReturnValue(
  of([
    {
      id: 'VMOperations',
      name: 'VMOperation Peak Duration',
      requiredEvents: {
        'jdk.ExecuteVMOperation': 'ENABLED',
      },
      topic: 'vm_operations',
    },
    {
      id: 'biasedLockingRevocation',
      name: 'Biased Locking Revocation',
      requiredEvents: {
        'jdk.BiasedLockClassRevocation': 'ENABLED',
      },
      topic: 'biased_locking',
    },
    {
      id: 'biasedLockingRevocationPause',
      name: 'Biased Locking Revocation Pauses',
      requiredEvents: {
        'jdk.ExecuteVMOperation': 'ENABLED',
      },
      topic: 'vm_operations',
    },
  ]),
);

describe('<TopologyConfig />', () => {
  afterEach(cleanup);

  it('renders correctly', async () => {
    const tree = await renderSnapshot({
      routerConfigs: {
        routes: [
          {
            path: '/settings/topology',
            element: React.createElement(TopologyConfig.content, null),
          },
        ],
      },
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });
});
