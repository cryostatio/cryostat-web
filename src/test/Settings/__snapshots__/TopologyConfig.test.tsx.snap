// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopologyConfig /> renders correctly 1`] = `
<div
  className="pf-v5-l-stack pf-m-gutter"
>
  <div
    className="pf-v5-l-stack__item"
  >
    <div
      className="pf-v5-c-helper-text"
    >
      <div
        className="pf-v5-c-helper-text__item"
      >
        <span
          className="pf-v5-c-helper-text__item-text"
        >
          The selected automated analysis report rule IDs will be ignored by the Topology view when determining the health status score of a target node and the number of rules that were evaluated to arrive at that score. Analysis results of active Targets or of archived Recordings will obey server analysis settings and ignore these selections.
        </span>
      </div>
    </div>
  </div>
  <div
    className="pf-v5-l-stack__item pf-m-fill"
  >
    <div
      className="pf-v5-c-dual-list-selector"
      id="pf-random-id-0"
    >
      <div
        className="pf-v5-c-dual-list-selector__pane pf-m-available"
      >
        <div
          className="pf-v5-c-dual-list-selector__header"
        >
          <div
            className="pf-v5-c-dual-list-selector__title"
          >
            <div
              className="pf-v5-c-dual-list-selector__title-text"
            >
              Analyzed rules
            </div>
          </div>
        </div>
        <div
          className="pf-v5-c-dual-list-selector__tools"
        >
          <div
            className="pf-v5-c-dual-list-selector__tools-filter"
          >
            <div
              className="pf-v5-c-text-input-group"
            >
              <div
                className="pf-v5-c-text-input-group__main pf-m-icon"
                onKeyDown={[Function]}
              >
                <span
                  className="pf-v5-c-text-input-group__text"
                >
                  <span
                    className="pf-v5-c-text-input-group__icon"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </span>
                  <input
                    aria-label="Available search input"
                    className="pf-v5-c-text-input-group__text-input"
                    disabled={false}
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          className="pf-v5-c-dual-list-selector__status"
        >
          <div
            className="pf-v5-c-dual-list-selector__status-text"
            id="pf-random-id-0-available-pane-status"
          >
            0 of 3 items selected
          </div>
        </div>
        <div
          className="pf-v5-c-dual-list-selector__menu"
          tabIndex={0}
        >
          <ul
            aria-activedescendant=""
            aria-labelledby="pf-random-id-0-available-pane-status"
            aria-multiselectable={true}
            className="pf-v5-c-dual-list-selector__list"
            role="tree"
          >
            <li
              aria-expanded="true"
              aria-selected={false}
              className="pf-v5-c-dual-list-selector__list-item pf-m-expandable pf-m-expanded"
              id="biased_locking"
              role="treeitem"
            >
              <div
                className="pf-v5-c-dual-list-selector__list-item-row pf-m-check"
              >
                <div
                  className="pf-v5-c-dual-list-selector__item"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  <span
                    className="pf-v5-c-dual-list-selector__item-main"
                  >
                    <div
                      className="pf-v5-c-dual-list-selector__item-toggle"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      tabIndex={-1}
                    >
                      <span
                        className="pf-v5-c-dual-list-selector__item-toggle-icon"
                      >
                        <svg
                          aria-hidden={true}
                          aria-labelledby={null}
                          className="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 256 512"
                          width="1em"
                        >
                          <path
                            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                          />
                        </svg>
                      </span>
                    </div>
                    <span
                      className="pf-v5-c-dual-list-selector__item-check"
                    >
                      <input
                        checked={false}
                        onChange={[Function]}
                        onClick={[Function]}
                        onKeyDown={[Function]}
                        tabIndex={-1}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="pf-v5-c-dual-list-selector__item-text"
                    >
                      biased_locking
                    </span>
                    <span
                      className="pf-v5-c-dual-list-selector__item-count"
                    >
                      <span
                        className="pf-v5-c-badge pf-m-unread"
                      >
                        1
                      </span>
                    </span>
                  </span>
                </div>
              </div>
              <ul
                className="pf-v5-c-dual-list-selector__list"
                role="group"
              >
                <li
                  aria-selected={false}
                  className="pf-v5-c-dual-list-selector__list-item"
                  id="biasedLockingRevocation"
                  role="treeitem"
                >
                  <div
                    className="pf-v5-c-dual-list-selector__list-item-row pf-m-check"
                  >
                    <div
                      className="pf-v5-c-dual-list-selector__item"
                      onClick={[Function]}
                      tabIndex={-1}
                    >
                      <span
                        className="pf-v5-c-dual-list-selector__item-main"
                      >
                        <span
                          className="pf-v5-c-dual-list-selector__item-check"
                        >
                          <input
                            checked={false}
                            onChange={[Function]}
                            onClick={[Function]}
                            onKeyDown={[Function]}
                            tabIndex={-1}
                            type="checkbox"
                          />
                        </span>
                        <span
                          className="pf-v5-c-dual-list-selector__item-text"
                        >
                          Biased Locking Revocation
                        </span>
                      </span>
                    </div>
                  </div>
                </li>
              </ul>
            </li>
            <li
              aria-expanded="true"
              aria-selected={false}
              className="pf-v5-c-dual-list-selector__list-item pf-m-expandable pf-m-expanded"
              id="vm_operations"
              role="treeitem"
            >
              <div
                className="pf-v5-c-dual-list-selector__list-item-row pf-m-check"
              >
                <div
                  className="pf-v5-c-dual-list-selector__item"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  <span
                    className="pf-v5-c-dual-list-selector__item-main"
                  >
                    <div
                      className="pf-v5-c-dual-list-selector__item-toggle"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      tabIndex={-1}
                    >
                      <span
                        className="pf-v5-c-dual-list-selector__item-toggle-icon"
                      >
                        <svg
                          aria-hidden={true}
                          aria-labelledby={null}
                          className="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 256 512"
                          width="1em"
                        >
                          <path
                            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                          />
                        </svg>
                      </span>
                    </div>
                    <span
                      className="pf-v5-c-dual-list-selector__item-check"
                    >
                      <input
                        checked={false}
                        onChange={[Function]}
                        onClick={[Function]}
                        onKeyDown={[Function]}
                        tabIndex={-1}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="pf-v5-c-dual-list-selector__item-text"
                    >
                      vm_operations
                    </span>
                    <span
                      className="pf-v5-c-dual-list-selector__item-count"
                    >
                      <span
                        className="pf-v5-c-badge pf-m-unread"
                      >
                        2
                      </span>
                    </span>
                  </span>
                </div>
              </div>
              <ul
                className="pf-v5-c-dual-list-selector__list"
                role="group"
              >
                <li
                  aria-selected={false}
                  className="pf-v5-c-dual-list-selector__list-item"
                  id="biasedLockingRevocationPause"
                  role="treeitem"
                >
                  <div
                    className="pf-v5-c-dual-list-selector__list-item-row pf-m-check"
                  >
                    <div
                      className="pf-v5-c-dual-list-selector__item"
                      onClick={[Function]}
                      tabIndex={-1}
                    >
                      <span
                        className="pf-v5-c-dual-list-selector__item-main"
                      >
                        <span
                          className="pf-v5-c-dual-list-selector__item-check"
                        >
                          <input
                            checked={false}
                            onChange={[Function]}
                            onClick={[Function]}
                            onKeyDown={[Function]}
                            tabIndex={-1}
                            type="checkbox"
                          />
                        </span>
                        <span
                          className="pf-v5-c-dual-list-selector__item-text"
                        >
                          Biased Locking Revocation Pauses
                        </span>
                      </span>
                    </div>
                  </div>
                </li>
                <li
                  aria-selected={false}
                  className="pf-v5-c-dual-list-selector__list-item"
                  id="VMOperations"
                  role="treeitem"
                >
                  <div
                    className="pf-v5-c-dual-list-selector__list-item-row pf-m-check"
                  >
                    <div
                      className="pf-v5-c-dual-list-selector__item"
                      onClick={[Function]}
                      tabIndex={-1}
                    >
                      <span
                        className="pf-v5-c-dual-list-selector__item-main"
                      >
                        <span
                          className="pf-v5-c-dual-list-selector__item-check"
                        >
                          <input
                            checked={false}
                            onChange={[Function]}
                            onClick={[Function]}
                            onKeyDown={[Function]}
                            tabIndex={-1}
                            type="checkbox"
                          />
                        </span>
                        <span
                          className="pf-v5-c-dual-list-selector__item-text"
                        >
                          VMOperation Peak Duration
                        </span>
                      </span>
                    </div>
                  </div>
                </li>
              </ul>
            </li>
          </ul>
        </div>
      </div>
      <div
        aria-label="Selector controls"
        className="pf-v5-c-dual-list-selector__controls"
        role="group"
        tabIndex={0}
      >
        <div
          className="pf-v5-c-dual-list-selector__controls-item"
        >
          <button
            aria-disabled={true}
            aria-label="Add selected"
            className="pf-v5-c-button pf-m-plain pf-m-disabled"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe={true}
            disabled={true}
            onClick={[Function]}
            role={null}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 256 512"
              width="1em"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
              />
            </svg>
          </button>
        </div>
        <div
          className="pf-v5-c-dual-list-selector__controls-item"
        >
          <button
            aria-disabled={false}
            aria-label="Add all"
            className="pf-v5-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-2"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe={true}
            disabled={false}
            onClick={[Function]}
            role={null}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
              />
            </svg>
          </button>
        </div>
        <div
          className="pf-v5-c-dual-list-selector__controls-item"
        >
          <button
            aria-disabled={true}
            aria-label="Remove all"
            className="pf-v5-c-button pf-m-plain pf-m-disabled"
            data-ouia-component-id="OUIA-Generated-Button-plain-3"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe={true}
            disabled={true}
            onClick={[Function]}
            role={null}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M223.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L319.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L393.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34zm-192 34l136 136c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9L127.9 256l96.4-96.4c9.4-9.4 9.4-24.6 0-33.9L201.7 103c-9.4-9.4-24.6-9.4-33.9 0l-136 136c-9.5 9.4-9.5 24.6-.1 34z"
              />
            </svg>
          </button>
        </div>
        <div
          className="pf-v5-c-dual-list-selector__controls-item"
        >
          <button
            aria-disabled={true}
            aria-label="Remove selected"
            className="pf-v5-c-button pf-m-plain pf-m-disabled"
            data-ouia-component-id="OUIA-Generated-Button-plain-4"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe={true}
            disabled={true}
            onClick={[Function]}
            role={null}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 256 512"
              width="1em"
            >
              <path
                d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        className="pf-v5-c-dual-list-selector__pane pf-m-chosen"
      >
        <div
          className="pf-v5-c-dual-list-selector__header"
        >
          <div
            className="pf-v5-c-dual-list-selector__title"
          >
            <div
              className="pf-v5-c-dual-list-selector__title-text"
            >
              Ignored rules
            </div>
          </div>
        </div>
        <div
          className="pf-v5-c-dual-list-selector__tools"
        >
          <div
            className="pf-v5-c-dual-list-selector__tools-filter"
          >
            <div
              className="pf-v5-c-text-input-group"
            >
              <div
                className="pf-v5-c-text-input-group__main pf-m-icon"
                onKeyDown={[Function]}
              >
                <span
                  className="pf-v5-c-text-input-group__text"
                >
                  <span
                    className="pf-v5-c-text-input-group__icon"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </span>
                  <input
                    aria-label="Chosen search input"
                    className="pf-v5-c-text-input-group__text-input"
                    disabled={false}
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          className="pf-v5-c-dual-list-selector__status"
        >
          <div
            className="pf-v5-c-dual-list-selector__status-text"
            id="pf-random-id-0-chosen-pane-status"
          >
            0 of 0 items selected
          </div>
        </div>
        <div
          className="pf-v5-c-dual-list-selector__menu"
          tabIndex={0}
        >
          <ul
            className="pf-v5-c-dual-list-selector__list"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="pf-v5-l-stack__item"
  >
    <button
      aria-disabled={false}
      aria-label={null}
      className="pf-v5-c-button pf-m-primary"
      data-ouia-component-id="OUIA-Generated-Button-primary-1"
      data-ouia-component-type="PF5/Button"
      data-ouia-safe={true}
      disabled={false}
      onClick={[Function]}
      role={null}
      type="button"
    >
      Restore Defaults
    </button>
  </div>
</div>
`;
